aql-engine/src/test/java/org/ehrbase/openehr/aqlengine/service/AqlQueryServiceImpTest.java

```diff
@@ -55,4 +101,51 @@ void resolveEhrCompositions(String srcAql, String expectedAql) {
         AqlQueryServiceImp.replaceEhrPaths(aqlQuery);
         assertThat(aqlQuery.render()).isEqualTo(expectedAql.replaceAll(" +", " "));
     }
+
+    @ParameterizedTest
+    @CsvSource(
+            textBlock =
+                    """
+                5||10||Query contains a LIMIT clause, fetch and offset parameters must not be used
+                5|20||40|Query contains a LIMIT clause, fetch and offset parameters must not be used
+                5|||30|Query contains a LIMIT clause, fetch and offset parameters must not be used
+                |||42|Query parameter for offset 42 provided without a fetch limit
+            """,
+            delimiterString = "|")
+    void queryOffsetLimitRejected(
+            String aqlLimit, String aqlOffset, String paramLimit, String paramOffset, String message) {
+
+        assertThatThrownBy(() -> runQueryTest(aqlLimit, aqlOffset, paramLimit, paramOffset))
+                .isInstanceOf(UnprocessableEntityException.class)
+                .hasMessage(message);
+    }
+
+    @ParameterizedTest
+    @CsvSource(
+            textBlock =
+                    """
+                5|||
+                5|15||
+                ||20|
+                ||20|25
+            """,
+            delimiterString = "|")
+    void queryOffsetLimitAccepted(String aqlLimit, String aqlOffset, String paramLimit, String paramOffset) {
+        runQueryTest(aqlLimit, aqlOffset, paramLimit, paramOffset);
+    }
+
+    private void runQueryTest(String aqlLimit, String aqlOffset, String paramLimit, String paramOffset) {
+        // @format:off
+        String query = "SELECT s FROM EHR_STATUS s %s %s".formatted(
+                Optional.ofNullable(aqlLimit).filter(s -> !s.isEmpty()).map(s -> "LIMIT " + s).orElse(""),
+                Optional.ofNullable(aqlOffset).filter(s -> !s.isEmpty()).map(s -> "OFFSET " + s).orElse("")
+        );
+        service().query(new AqlQueryRequest(
```

Comment:
If ::buildAqlQuery was package protected, the static method could be tested without mocking everything.
If the parameter replacement was extracted into i´ts own method, it could be tested with even fever side effects.

Reply:
adjusted to test only `buildAqlQuery`